pyquil/api/_qam.py

```diff
@@ -41,16 +42,20 @@ class QAM(ABC):
     pretend to be a QPI-compliant quantum computer.
     """
 
+    status: str
+    _client: Client
+    _variables_shim: Dict[ParameterAref, Union[int, float]]
+    executable: Optional[QuantumExecutable]
+    experiment: Optional[Experiment] = None
     _memory_results: Dict[str, np.ndarray]
 
     @_record_call
-    def __init__(self) -> None:
+    def __init__(self, client: Optional[Client] = None) -> None:
+        self._client = client or Client()
```

Comment:
This seems to be pretty idiomatic. I think the factory patter could serve us well.